go/parquet/compress/compress_test.go

```diff
@@ -0,0 +1,128 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compress_test
+
+import (
+	"bytes"
+	"io/ioutil"
+	"math/rand"
+	"testing"
+
+	"github.com/apache/arrow/go/parquet/compress"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	RandomDataSize       = 3 * 1024 * 1024
+	CompressibleDataSize = 8 * 1024 * 1024
+)
+
+func makeRandomData(size int) []byte {
+	ret := make([]byte, size)
+	r := rand.New(rand.NewSource(1234))
+	r.Read(ret)
+	return ret
+}
+
+func makeCompressibleData(size int) []byte {
+	const base = "Apache Arrow is a cross-language development platform for in-memory data"
+
+	data := make([]byte, size)
+	n := copy(data, base)
+	for i := n; i < len(data); i *= 2 {
+		copy(data[i:], data[:i])
+	}
+	return data
+}
+
+func TestCompressDataOneShot(t *testing.T) {
+	tests := []struct {
+		c compress.Compression
+	}{
+		{compress.Codecs.Uncompressed},
+		{compress.Codecs.Snappy},
+		{compress.Codecs.Gzip},
+		// {compress.Codecs.Lzo},
+		{compress.Codecs.Brotli},
+		{compress.Codecs.Lz4},
+		{compress.Codecs.Zstd},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.c.String(), func(t *testing.T) {
+			codec := compress.GetCodec(tt.c)
+			data := makeCompressibleData(CompressibleDataSize)
+
+			buf := make([]byte, codec.CompressBound(int64(len(data))))
+			compressed := codec.Encode(buf, data)
+			assert.Same(t, &buf[0], &compressed[0])
+
+			out := make([]byte, len(data))
+			uncompressed := codec.Decode(out, compressed)
+			assert.Same(t, &out[0], &uncompressed[0])
+
+			assert.Exactly(t, data, uncompressed)
+		})
+	}
+}
+
+func TestCompressReaderWriter(t *testing.T) {
+	tests := []struct {
+		c compress.Compression
+	}{
+		{compress.Codecs.Uncompressed},
+		{compress.Codecs.Snappy},
+		{compress.Codecs.Gzip},
+		// {compress.Codecs.Lzo},
+		{compress.Codecs.Brotli},
+		{compress.Codecs.Lz4},
+		{compress.Codecs.Zstd},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.c.String(), func(t *testing.T) {
+			var buf bytes.Buffer
+			codec := compress.GetCodec(tt.c)
+			data := makeRandomData(RandomDataSize)
+
+			wr := codec.NewWriter(&buf)
+
+			const chunkSize = 1111
```

Comment:
pulled this test from the C++ implementation tests. Ultimately it's because it's a number that is small enough to make sure we'll have multiple chunks but large enough that it'll have some compression it can do :)

Reply:
I'm surprised C++ doesn't use a power of 2 (1024?)

Reply:
I think the intent was to ensure that there is a chunk at the end which is not a full chunk to make sure we test that handling properly.